### PR TITLE
Use the correct authority to validate members of groups in the admin pages

### PR DESCRIPTION
**Pull request by jon-betts**
* Previously members were always validated against the default authority
 * They are now validated against authority matching the organisation
 * This involved moving the member validation to the global validation method

### CODE REVIEW

File: tests/h/schemas/forms/admin/group_test.py [hunk 1]
```diff
@@ -162,18 +183,27 @@ def group_data(factories):
...
-def user_svc(pyramid_config):
+def user_svc(pyramid_config, factories):
     svc = mock.create_autospec(UserService, spec_set=True, instance=True)
     pyramid_config.register_service(svc, name="user")
+
+    def fetch(username, authority):
+        if "invalid" in username:
+            return False
+
+        return factories.User()
```
Comment (jon-betts) [inline tests/h/schemas/forms/admin/group_test.py:203]: Having this fail everytime caused non-isolated tests where multiple things were failing at the same time. Our tests should now always pass unless we set a username with "invalid" in it.

File: h/schemas/forms/admin/group.py [hunk 2]
```diff
@@ -24,56 +22,65 @@
...
+    The creator and members of a group must belong to the same authority as the
+    group and the group's organization.
+    """
+    user_svc = form.bindings["user_svc"]
 
-        """
-        user_svc = kw["user_svc"]
+    # Get the authority from the list of organisations based on the one
+    # actually picked in the form
+    authority = form.bindings["organizations"][value["organization"]].authority
```
Comment (jon-betts) [inline h/schemas/forms/admin/group.py:36]: There was some highly over commented code here. I've packed this down to a single line

File: h/schemas/forms/admin/group.py [hunk 3]
```diff
@@ -24,56 +22,65 @@
 VALID_GROUP_TYPES = (("restricted", _("Restricted")), ("open", _("Open")))
 
 
-@colander.deferred
-def group_creator_validator(node, kw):
```
Comment (jon-betts) [inline h/schemas/forms/admin/group.py:28]: I think `colander.deferred` was only being used here to access `kw`, but the values here are available as `form.bindings` which is also much more informative. This matches the other parts where we don't use `deferred` to access these variables.

File: h/schemas/forms/admin/group.py [hunk 4]
```diff
@@ -24,56 +22,65 @@
...
+        if not user_svc.fetch(node_value, authority):
+            # If we don't attach errors in the right position, they don't
+            # appear. For a list type value the correct position is the index
+            # of the value in the list
+            members_error.add(
+                colander.Invalid(
+                    member_node,
+                    _("User '{username}' not found at authority '{authority}'").format(
+                        username=node_value, authority=authority
+                    ),
```
Comment (jon-betts) [inline h/schemas/forms/admin/group.py:65]: We now use the same error message for members as the creator, which includes the authority.

File: h/schemas/forms/admin/group.py [hunk 5]
```diff
@@ -24,56 +22,65 @@
 VALID_GROUP_TYPES = (("restricted", _("Restricted")), ("open", _("Open")))
 
 
-@colander.deferred
-def group_creator_validator(node, kw):
-    def validate(form, value):
-        """
-        Validate that the creator username exists in the organization's authority.
+def username_validator(form, value):
```
Comment (lyzadanger) [inline h/schemas/forms/admin/group.py:25]: This function is used as a form-/schema-level validator (versus a single node validator). Does that warrant a comment or a rename?

File: h/schemas/forms/admin/group.py (hunk 2 above)
Comment (lyzadanger) [reply h/schemas/forms/admin/group.py:36]: I assure you that was me, even if I check history and it wasn't...seems like something I would do.

File: h/schemas/forms/admin/group.py [hunk 6]
```diff
@@ -24,56 +22,65 @@
...
+                    ),
+                ),
+                pos=pos,
+            )
 
-    return validate
+    if members_error.children:
+        # For non-lists the correct position is the index within the flat list
+        # of children that colander stores internally for the parent node
+        parent_exc.add(members_error, pos=form.children.index(members_node))
```
Comment (lyzadanger) [inline h/schemas/forms/admin/group.py:73]: By this point, it's pretty easy to lose track of what `non-list` is in context here: `parent_exc` is, at least in _practice_ here, any error(s) associated with the form itself? There's about one level of abstraction too many for my little beady brain's comfort here, but it's not catastrophic.

File: h/schemas/forms/admin/group.py [hunk 7]
```diff
@@ -119,7 +126,7 @@ def group_organization_select_widget(node, kw):
 class CreateAdminGroupSchema(CSRFSchema):
```
Comment (lyzadanger) [inline h/schemas/forms/admin/group.py:126]: At this point, are both the create and edit forms using this schema? Is a rename or comment in order? Or is it obvious enough?

File: h/schemas/forms/admin/group.py (hunk 6 above)
Comment (jon-betts) [reply h/schemas/forms/admin/group.py:73]: No snappy way I can think of to explain this...

From what I understand the structure of the errors you produce must _mostly_ reflect the structure of the schema. The issue is that the schema has a single node representing the structure of all entries in a list. So the schema be like:

```
root
    creator (pos:0)
    members (pos:1)
        member (pos:0)  # represents many, but is a single thing
```

Where as the data is like:

```
root
    creator (pos:0)
    members (pos:1)
          member.0 (pos:0)
          member.1 (pos:1)
          member.2 (pos:2)
```

To add an error about the creator we need a root error, which we then add an error to in the 0th position, reflecting it's place in the schema:

```
root = Invalid(form_node, None)
creator = Invalid(creator_node, "blah") 

root.add(creator, 0)  # 0 is the position in the schema
```

To add an error about a specific member it's different. The first part is the same, we need to nest a members parent error to reflect the position of the members node in the schema, but we then switch to using the value position in the data for the many values for member:

```
root = Invalid(form_node, None)
members = Invalid(members_node, None)
member = Invalid(member_node, "Wrong username")

root.add(members, 1)  # 1 is the position in the schema
members.add(member, 2) # 2 here is the position in the values
```

----

Long story short, both are individual errors about a part of the form, but one is an error about a single valued field: the other is an error about one value within a multi valued list field.

File: h/schemas/forms/admin/group.py (hunk 5 above)
Comment (jon-betts) [reply h/schemas/forms/admin/group.py:25]: This reflects how it was named before. It was always a form level validator, but it's behavior was only to validate the creator. So it was called: `group_creator_validator`. Now it's doing all usernames I've broadened it to `username_validator`.

I think it's better to describe what it's trying to do, rather than how Colander forces us to do it.

File: h/schemas/forms/admin/group.py (hunk 7 above)
Comment (jon-betts) [reply h/schemas/forms/admin/group.py:126]: This is a pretty fair point. it's probably just the `AdminGroupSchema` if it's used equally for both.

File: h/schemas/forms/admin/group.py (hunk 5 above)
Comment (jon-betts) [reply h/schemas/forms/admin/group.py:25]: Added a comment now

File: h/schemas/forms/admin/group.py (hunk 7 above)
Comment (jon-betts) [reply h/schemas/forms/admin/group.py]: This has been renamed to just "AdminGroupSchema" now
